Ranit/emulator init

Kawai-Senpai opened the pull request:
This pull request introduces an emulator mode for the Flight Matrix Simulation Software, adds new configuration options, and includes various improvements to the codebase. The most important changes are grouped into documentation updates, configuration changes, and code enhancements.

### Documentation Updates:
* [`README.md`](diffhunk://#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R19-R93): Added a new section detailing the emulator mode, including data format requirements, launching instructions, and enabling debug mode.

### Configuration Changes:
* [`config/config.yaml`](diffhunk://#diff-38df760413887d0cc4ee5707919e6390456a863d326f6080644cc3d9c72e4bbaR1-R8): Updated configuration parameters, including setting `timer_delay` to `0.0` and adding a placeholder parameter for the `drone_controller`. [[1]](diffhunk://#diff-38df760413887d0cc4ee5707919e6390456a863d326f6080644cc3d9c72e4bbaR1-R8) [[2]](diffhunk://#diff-38df760413887d0cc4ee5707919e6390456a863d326f6080644cc3d9c72e4bbaR17-R23)
* [`config/config_emulator.yaml`](diffhunk://#diff-21e1793c0b7627189ea659894e2a8bb375219d18ff296dce93e239a97330c481R1-R19): Added a new configuration file for the emulator mode with detailed parameters for resolution, publishers, and timer settings.

### Code Enhancements:
* [`flightmatrix_ros2/flightmatrix_publisher.py`](diffhunk://#diff-05ade861789d702f1c4f247e6703338d8505d05cd4b307b7ddbf5df3b5883d4cL12-L17): Added logging for configuration loading, introduced a `safe_exit` method for handling critical errors, and improved shared memory access error handling. [[1]](diffhunk://#diff-05ade861789d702f1c4f247e6703338d8505d05cd4b307b7ddbf5df3b5883d4cL12-L17) [[2]](diffhunk://#diff-05ade861789d702f1c4f247e6703338d8505d05cd4b307b7ddbf5df3b5883d4cR31-R36) [[3]](diffhunk://#diff-05ade861789d702f1c4f247e6703338d8505d05cd4b307b7ddbf5df3b5883d4cR99-R116) [[4]](diffhunk://#diff-05ade861789d702f1c4f247e6703338d8505d05cd4b307b7ddbf5df3b5883d4cL110-R128)
* [`flightmatrix_ros2/flightmatrix_publisher_emulator.py`](diffhunk://#diff-bf31dcb3b9b9ffe42e1823e14aefabd91ac735e9675bfb25bd3605aa39b74d77R1-R346): Implemented the emulator functionality, including data loading, frame publishing, sensor data publishing, and synchronization monitoring.
* [`launch/flightmatrix.launch.py`](diffhunk://#diff-94075ef66c256f7031100dae22861324e4bbf27adc0ca58f3907c981e416b9d7L1-R9): Simplified the configuration file path setup by removing unnecessary imports and hardcoding the config file path.

Kawai-Senpai commented:
Didn't use the git message. Will from the next time. But just to summarise, I added the emulator publisher which publish and emulate all the frames and sensor datas. It takes the recorded data as the fuel which it expects to be in a certain format (detailed out in readme). You can use the sample ones in the drive to just plug and play. Tested with Rviz.

shandilya1998 commented on README.md:
```diff
@@ -10,51 +10,175 @@ For more information, visit the [GitHub repository](https://github.com/Kawai-Sen
-    share_dir = get_package_share_directory('flightmatrix_ros2')
-    config_file = LaunchConfiguration('config_file')
+    
+    config_file_path = "/path/to/config.yaml"
```
Can you specify a path to a file that can be commit here? for instance "/ws/ros_ws/src/flight_matrix_bridge/config/...". This will ensure that atleast for the toolkit, we may need not specify the path, anyone else using the package will have to.

Kawai-Senpai commented:
This pull request introduces several significant updates to the Flight Matrix ROS2 nodes, including documentation enhancements, the addition of an emulator mode, and configuration file changes. The most important changes are summarized below:

### Documentation Enhancements:
* Added a "Last Updated" section and updated the image link in `README.md`.
* Included detailed instructions and data format requirements for the new emulator mode in `README.md`.

### Emulator Mode:
* Introduced an emulator mode in `README.md`, which allows replaying recorded data without the actual Flight Matrix software.
* Added `flightmatrix_emulator.launch.py` to start the emulator node for replaying recorded data.

### Configuration File Changes:
* Updated `config.yaml` to include `drone_controller` parameters and set `sensor_data` to true. [[1]](diffhunk://#diff-38df760413887d0cc4ee5707919e6390456a863d326f6080644cc3d9c72e4bbaR1-R8) [[2]](diffhunk://#diff-38df760413887d0cc4ee5707919e6390456a863d326f6080644cc3d9c72e4bbaR17-R23)
* Created `config_emulator.yaml` for emulator mode settings, including data paths and projection settings.

### Code Enhancements:
* Improved error handling and logging in `flightmatrix_publisher.py`, including a new `safe_exit` method for safe program termination. [[1]](diffhunk://#diff-05ade861789d702f1c4f247e6703338d8505d05cd4b307b7ddbf5df3b5883d4cR31-R36) [[2]](diffhunk://#diff-05ade861789d702f1c4f247e6703338d8505d05cd4b307b7ddbf5df3b5883d4cR99-R116) [[3]](diffhunk://#diff-05ade861789d702f1c4f247e6703338d8505d05cd4b307b7ddbf5df3b5883d4cL110-R128)
* Removed unused imports from `flightmatrix_publisher.py`.

### Launch Files:
* Updated `flightmatrix.launch.py` to use a fixed path for the configuration file.
* Added `flightmatrix_emulator.launch.py` to the `setup.py` for proper installation and execution. [[1]](diffhunk://#diff-fdaf58df454d083c915796bbbcd29cbdfdea3831cedb58ae0897ffc476bae4a6R1-R30) [[2]](diffhunk://#diff-60f61ab7a8d1910d86d9fda2261620314edcae5894d5aaa236b821c7256badd7R15-R16) [[3]](diffhunk://#diff-60f61ab7a8d1910d86d9fda2261620314edcae5894d5aaa236b821c7256badd7R30)

Kawai-Senpai commented:
@shandilya1998 Just checked, I still can't see your comments.
